DIPs/DIPNNNN.md

```diff
@@ -0,0 +1,295 @@
+
+```D
+class Base
+{
+}
+```
+
+At some point maintainer of the library wants to add a new method to this
+class, `void method()`, but doing so risks breaking user code too:
+
+```D
+class Derived : Base
+{
+    void method(bool) { }
+}
+```
+
+If method with same name was already present in derived class, it will error out
+because of either missing `override` keyword or attempt on non-covariant override.
+
+### Existing solutions
+
+For many libraries both mentioned problems are of very low importance. These are
+widely known properties of D symbol resolution and normally projects upgrading
+their dependencies to new feature release do expect some minor maintenance
+effort.
+
+However, there are also certain libraries that set an extremely high stability
+expectations, with most notable example being [druntime](https://github.com/dlang/druntime),
+D standard runtime library. Even very minor breakage coming from such project
+can affect many users and is likely to come unexpected.
+
+In such projects, there are few existing approaches to reduce the negative
+impact of adding new symbols:
+
+#### Using more obscure name
+
+Most common approach is to simply pick names that are less likely to be already
+used in downstream projects. It is simple and practical solution that doesn't
+require any language support but it has obvious scalability issues:
+
+- Results in unnecessarily baroque names, harming learning curve
+- Picking unused names becomes harder as amount of downstream projects grow
+- Adding new symbol has high chance of being blocked on naming debate
+
+#### Emulating namespaces
+
+Another form of name obfuscation is using "namespace" aggregates:
+
+```D
+final abstract class Foo
+{
+    static:
+
+    void foo() {};
+}
+```
+
+It makes naming clashes even less likely and resulting names more obvious
+but suffers from different important issues:
+
+- Can't be used for new symbols only, fundamental library API design decision
+- Does not leverage D module system
+- Can't be use to solve [Problem 2](#problem-2-override-clash)
+
+#### Deprecations
+
+Contrary to adding new symbol, removing or renaming one is much easier task
+because D supports `deprecated` keyword, ensuring that users of the library
+will get early notification about to-be-removed symbol before it is actually
+gone.
+
+Similar warning should also be suitable for adding new symbols in critical
+libraries but language currently lack a way to express such semantics.
+
+### Proposal
+
+Introduce new symbol kind concept into compiler, *future symbol*. This term does
+not refer to any language syntax and is intended to represent new concept in
+compiler internals. It has to have special treatment during semantic analysis:
+
+1. If any symbol resolution ends up in *future symbol* being present among
+   valid candidates, deprecation message has to be printed and symbol
+   resolution continues after that as if it didn't exist.
```

Comment:
Valid candidates is fairly unprecise. IMO it should only be an error if it's actually picked after hijack protection and overload resolution, otherwise it's practically not a conflicting symbol.
For example introducing a top level `front` method for a RedBlackTree wouldn't conflict with any `front` in std.array.